internal/pgkeyvalue/pgkeyvalue.go

```diff
@@ -86,6 +108,76 @@ func (c *PGKeyValueStore) createTable(ctx context.Context) error {
 	return err
 }
 
+func (c *PGKeyValueStore) addBatch(ctx context.Context, batch []*KeyValue) (map[string][]byte, error) {
+
+	addedByKey := map[string][]byte{}
+	keysToLookup := make([]string, 0, len(batch))
+	keysFound := 0
+
+	// first check the cache to see if we have added anything
+	for _, kv := range batch {
+		if val, ok := c.cache.Get(kv.Key); ok {
+			addedByKey[kv.Key] = val.([]byte)
+			keysFound++
+		} else {
+			keysToLookup = append(keysToLookup, kv.Key)
+		}
+	}
+
+	if keysFound == len(batch) {
+		return addedByKey, nil
+	}
+
+	err := c.db.BeginTxFunc(ctx, pgx.TxOptions{}, func(tx pgx.Tx) error {
+
+		// See which keys are already in the db
+		sql := fmt.Sprintf("SELECT key, value from %s WHERE key = ANY($1)", c.tableName)
```

Comment:
Are there any concerns about concurrent calls updating the same ids?

Reply:
You're right- it's not as bulletproof as it could be- I've modified it to be a single statement which should make it work in all circumstances.

Note that the existing `Add` method suffers from the same issue as I'd just copied the pattern used there.  For now that's not an issue as we don't use it anywhere but we should fix in a follow up PR.